chain/network/src/peer_manager/tests/routing.rs

```diff
@@ -10,14 +12,136 @@ use crate::store;
 use crate::tcp;
 use crate::testonly::{make_rng, Rng};
 use crate::time;
+use crate::types::PeerInfo;
 use crate::types::PeerMessage;
 use near_o11y::testonly::init_test_logger;
 use near_store::db::TestDB;
 use pretty_assertions::assert_eq;
 use rand::Rng as _;
 use std::collections::HashSet;
+use std::net::Ipv4Addr;
 use std::sync::Arc;
 
+fn make_configs(
+    chain: &data::Chain,
+    rng: &mut Rng,
+    num_nodes: usize,
+    num_boot_nodes: usize,
+    enable_outbound: bool,
+) -> Vec<NetworkConfig> {
+    let mut cfgs: Vec<_> = (0..num_nodes).map(|_i| chain.make_config(rng)).collect();
+    let boot_nodes: Vec<_> = cfgs[0..num_boot_nodes]
+        .iter()
+        .map(|c| PeerInfo { id: c.node_id(), addr: c.node_addr, account_id: None })
+        .collect();
+    for config in cfgs.iter_mut() {
+        config.outbound_disabled = !enable_outbound;
+        config.peer_store.boot_nodes = boot_nodes.clone();
+    }
+    cfgs
+}
+
+// test bootstrapping a two-node network with one boot node
+#[tokio::test]
+async fn from_boot_nodes() {
+    init_test_logger();
+    let mut rng = make_rng(921853233);
+    let rng = &mut rng;
+    let mut clock = time::FakeClock::default();
+    let chain = Arc::new(data::Chain::make(&mut clock, rng, 10));
+
+    tracing::info!(target:"test", "start two nodes");
+    let cfgs = make_configs(&chain, rng, 2, 1, true);
+    let pm0 = start_pm(clock.clock(), TestDB::new(), cfgs[0].clone(), chain.clone()).await;
+    let pm1 = start_pm(clock.clock(), TestDB::new(), cfgs[1].clone(), chain.clone()).await;
+
+    let id0 = pm0.cfg.node_id();
+    let id1 = pm1.cfg.node_id();
+
+    tracing::info!(target:"test", "wait for {id0} routing table");
+    pm0.wait_for_routing_table(&[(id1.clone(), vec![id1.clone()])]).await;
+    tracing::info!(target:"test", "wait for {id1} routing table");
+    pm1.wait_for_routing_table(&[(id0.clone(), vec![id0.clone()])]).await;
+}
+
+// test node 0 blacklisting node 1
+#[tokio::test]
+async fn blacklist_01() {
+    init_test_logger();
+    let mut rng = make_rng(921853233);
+    let rng = &mut rng;
+    let mut clock = time::FakeClock::default();
+    let chain = Arc::new(data::Chain::make(&mut clock, rng, 10));
+
+    tracing::info!(target:"test", "start two nodes with 0 blacklisting 1");
+    let mut cfgs = make_configs(&chain, rng, 2, 2, true);
+    cfgs[0].peer_store.blacklist =
+        [blacklist::Entry::from_addr(cfgs[1].node_addr.unwrap())].into_iter().collect();
+
+    let pm0 = start_pm(clock.clock(), TestDB::new(), cfgs[0].clone(), chain.clone()).await;
+    let pm1 = start_pm(clock.clock(), TestDB::new(), cfgs[1].clone(), chain.clone()).await;
+
+    let id0 = pm0.cfg.node_id();
+    let id1 = pm1.cfg.node_id();
+
+    tracing::info!(target:"test", "wait for {id0} routing table");
```

Comment:
This test is not very useful in this version:
wait_for_routing_table here will return immediately even before the nodes would even try to connect to each other (no matter what the banning status is). Instead you can force the banned node to try to connect to the other node and await for the connection to be rejected for a specific reason. However also please check whether we don't have such a test already.

Reply:
same for other tests here

Reply:
Thanks for the catch. I pushed a fix for the `blacklist_01` test which awaits a ConnectionClosed event before checking the routing tables. I also verified that no such event occurs if we do not blacklist anything.

Currently the test simply awaits any ConnectionClosed. Ideally it would await specifically a ConnectionClosed event with closing reason `ClosingReason::RejectedByPeerManager(RegisterPeerError::Blacklisted)`. However, we would need to expose some crate-private types (for example, `RegisterPeerError` from `chain/network/src/private_actix.rs`) to make that possible. Is it worth doing so?

Reply:
you should have access to RegisterPeerError from the test, since both reside in the same crate: "chain/network", no? Note that it is "crate-private", not "module-private".

Reply:
Ok, got it, I just needed to make my `wait_for_connection_closed` helper also crate-private.